csrc/scheduler/matmul_utils.cpp

```diff
@@ -294,6 +315,118 @@ std::string isMatmulFusionDefinitionSupported(
   return "";
 }
 
+// Assume that tens has a contiguous dimension, and that we will load rows of
+// that dimension, without merging with another dimension first. Then determine
+// the maximum vectorization that can be used.
+//
+// If the argument has no contiguous dimensions, then a vectorization width of 1
+// is returned.
+//
+// These rows can start at any multiple of the non-contiguous strides, so we
+// seek the largest power of 2 that divides all those other dimensions (capped
+// to 16) as well as the data pointer.
+int64_t maxRowVectorization(TensorView* tv, const at::Tensor& tens) {
+  const int64_t data_ptr_int =
+      static_cast<int64_t>(reinterpret_cast<std::uintptr_t>(tens.data_ptr()));
+  int64_t vec_size = scheduler_utils::maxVectorizationWidth(data_ptr_int);
+  vec_size = std::min(vec_size, 16l);
+  vec_size /= (int64_t)tens.element_size();
+  vec_size = std::max(vec_size, 1l);
+  if (vec_size == 1l) {
+    return vec_size;
+  }
+
+  NVF_ERROR(tens.ndimension() == tv->nDims());
+
+  // Find innermost dimension
+  bool contiguous = false;
+  for (int64_t i : c10::irange(tens.ndimension())) {
+    int64_t stride = tens.stride(i);
+    int64_t size = tens.size(i);
+    if (stride == 1) {
+      std::optional<bool> contig = tv->getContiguity().at(i);
+      if (contig.has_value() && !contig.value()) {
+        // If TensorView is marked discontiguous in inner dimension, we cannot
+        // vectorize regardless of input.
+        return 1l;
+      }
+      // Innermost dimension
+      contiguous = true;
+      vec_size =
+          std::min(vec_size, scheduler_utils::maxVectorizationWidth(size));
+    } else {
```

Comment:
Are you assuming rfactor domain == allocation domain? I don't think this works with @protonu's allocation domain support. I think we should check `contiguity[-1]` and `allocation_domain[-1]`'s extent?

Reply:
This code is working on the `at::Tensor` input that comes from expression evaluator, so we just look at sizes and strides.  I had in mind that these would be fusion inputs, but that might not always be the case in a segmented fusion, so this means we'd actually be evaluating preceding segments with the expression evaluator to get that tensor.

Instead, I suppose I should use `SchedulerRuntimeInfo::input_discontig_strides_` and evaluate the sizes. That also has the limitation that we will only have strides of input tensors.

Reply:
The allocation domain should match `tens.sizes()` and contiguity should reflect that and `tens.strides()` shouldn't it? I am not referring to rfactor or leaf domains anywhere here.

Reply:
> Instead, I suppose I should use `SchedulerRuntimeInfo::input_discontig_strides_` and evaluate the sizes. That also has the limitation that we will only have strides of input tensors.

I actually checked this, unfortunately I don't think it works. Due to the nature of how we schedule, we should effectively be only checking the innermost dim. For example, if the tensor has shape `(1024, 3)`, we can not vectorize, but IIUC `input_discontig_strides_` will give vectorization `8` because the `1024` is contiguous to the `3`.

Reply:
> This is scary.

I agree.

> Should we make `ExpressionEvaluator::getValue` public so that we can avoid using `evaluate` if we just want to get an existing value? This can be in a separate PR.

That's one way to go. It might be simpler though to just expose `input_discontig_strides_` with an accessor so that we can use it here, along with using `ExpressionEvaluator` on the allocation domain extents to get the sizes. That functionality does not use the expression evaluator at all so it is safe; it records strides directly in the `SchedulerRuntimeInfo` ctor. I think I can fit that change into this PR and it will make things much cleaner.

We also need to think about what happens when we have a segmented fusion. I suppose when a `TensorView` is not a complete Fusion input then we should have a special case where we assume its allocation domain matches its rfactor.

Reply:
I see what you're saying.

So for example given an `at::Tensor x` with
```
x.sizes() = 2 3 4
x.strides() = 15 5 1
```
This is equivalent in nvFuser to `TensorView* tv0` with
```
tv0->getMaybeRFactorDomain() = [ iS0{2} iS1{3} iS2{4} ]
tv0->getMaybeAllocationDomain() = [ iS0{2} iS1{3} iS2{4} ]
tv0->getContiguity() = [ t f t ]
```
Now if we have an `at::Tensor y` with
```
y.sizes() = 2 3 4
y.strides() = 16 1 4
```
then this is equivalent to `TensorView* tv1` with
```
tv1->getMaybeRFactorDomain() = [ iS0{2} iS1{3} iS2{4} ]
tv1->getMaybeAllocationDomain() = [ iS0{2} iS2{4} iS1{3} ]
tv1->getContiguity() = [ t f t ]
```
We see that `y.strides()` and `tv1->getContiguity()` do not match.

For this utility I don't really need the strides to be ordered as long as they are in the same order as sizes. But I think determining whether the Fusion has the dimension marked as contiguous might require a utility similar to the one you posted.

Reply:
@zasdfgbnm if I evaluate the sizes of the non-reduction IterDomains in `tv->getMaybeAllocationDomain()` and then I reverse sort the strides from the `at::Tensor`, won't those always correspond to one another? That means the only change I'd need to make is to add the stride sorting step.

Reply:
> @zasdfgbnm if I evaluate the sizes of the non-reduction IterDomains in `tv->getMaybeAllocationDomain()` and then I reverse sort the strides from the `at::Tensor`, won't those orders always correspond to one another and to that of `tv->getContiguity()`? That means the only change I'd need to make is to add the stride sorting step.

I think the only way that is guaranteed to be correct is to look at the permutation relationship of `getMaybeAllocationDomain` and `getMaybeRFactorDomain`. Looking at numbers does not always provide correct information. For example, for the same tensor of shape `(4, 5)` and stride `(5, 1)`, it can be interpreted as two different options:
```
allocation domain: I0{4}, I1{5}
contiguity: T, T
```
and
```
allocation domain: I1{5}, I0{4}
contiguity: F, F
```
Both are correct, although the resulting performance can be significantly different. Our code should handle both cases.

Reply:
Ah I see. Yeah you are right that my logic only works out if the tensor is contiguous. I'll change it!